### PR TITLE
Migrate security/pausable

### PR DESCRIPTION
**Pull request by andrew-fleming**
Fix (partial) #550.

#### PR Checklist

- [x] Tests
- [ ] Tried the feature on a public network
- [ ] Documentation

Supersedes #589.

### CODE REVIEW

File: src/openzeppelin/security/pausable.cairo [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    struct Storage {
+        _paused: bool,
+    }
+
+    #[event]
+    fn Paused(account: ContractAddress) {}
+
+    #[event]
+    fn Unpaused(account: ContractAddress) {}
```
Comment (Amxx) [inline src/openzeppelin/security/pausable.cairo:14]: Do we want to emit events here? In the solidity contracts, the "pausable" module is "silent".

Comment (martriay) [reply src/openzeppelin/security/pausable.cairo:14]: [is it](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/Pausable.sol#L89-L105)?

File: src/openzeppelin/mocks.cairo
Comment (martriay) [inline src/openzeppelin/mocks.cairo:1]: should this be under `tests/`?

File: src/openzeppelin/security/pausable.cairo (hunk 1 above)
Comment (Amxx) [reply src/openzeppelin/security/pausable.cairo:14]: My bad :/

File: src/openzeppelin/mocks/mock_pausable.cairo [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#[contract]
+mod MockPausable {
+    use openzeppelin::security::pausable::Pausable;
+
+    use starknet::ContractAddress;
+    use starknet::get_caller_address;
```
Comment (Amxx) [inline src/openzeppelin/mocks/mock_pausable.cairo:6]: Are these needed here?

**Review by Amxx — APPROVED** [review]
LGTM

File: src/openzeppelin/security/pausable.cairo [hunk 3]
```diff
@@ -0,0 +1,39 @@
+#[contract]
+mod Pausable {
+    use starknet::ContractAddress;
+    use starknet::get_caller_address;
+
+    struct Storage {
+        _paused: bool,
```
Comment (martriay) [inline src/openzeppelin/security/pausable.cairo:7]:
```suggestion
        paused: bool,
```
any reason not to?

File: src/openzeppelin/mocks/mock_pausable.cairo (hunk 2 above)
Comment (martriay) [inline src/openzeppelin/mocks/mock_pausable.cairo:6]: why are these needed?

File: src/openzeppelin/tests/test_pausable.cairo [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+#[should_panic(expected = ('Pausable: paused', ))]
+fn test_pause_when_paused() {
+    MockPausable::pause();
+    MockPausable::pause();
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('Pausable: paused', ))]
+fn test_pause_assertion() {
```
Comment (martriay) [inline src/openzeppelin/tests/test_pausable.cairo:25]:
```suggestion
fn test_pause_increment() {
```

File: src/openzeppelin/tests/test_pausable.cairo [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    MockPausable::pause();
+    MockPausable::assert_unpaused_and_increment();
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_unpause_when_paused() {
+    MockPausable::pause();
+    assert(MockPausable::is_paused() == true, 'Should be paused');
+    MockPausable::unpause();
```
Comment (martriay) [inline src/openzeppelin/tests/test_pausable.cairo:35]: should we explicitly test `is_paused()` is `false` after this?

File: src/openzeppelin/tests/test_pausable.cairo [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+fn test_pause_assertion() {
+    MockPausable::pause();
+    MockPausable::assert_unpaused_and_increment();
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_unpause_when_paused() {
+    MockPausable::pause();
+    assert(MockPausable::is_paused() == true, 'Should be paused');
```
Comment (martriay) [inline src/openzeppelin/tests/test_pausable.cairo:34]:
```suggestion
    assert(MockPausable::is_paused(), 'Should be paused');
```

File: src/openzeppelin/tests/test_pausable.cairo [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+#[test]
+#[available_gas(2000000)]
+fn test_pause_when_unpaused() {
+    assert(MockPausable::is_paused() == false, 'Should not be paused');
+    assert(MockPausable::get_count() == 0, 'Should be 0');
+    MockPausable::assert_unpaused_and_increment();
+    assert(MockPausable::get_count() == 1, 'Should increment');
+    MockPausable::pause();
+    assert(MockPausable::is_paused() == true, 'Should be paused');
```
Comment (martriay) [inline src/openzeppelin/tests/test_pausable.cairo:11]:
```suggestion
    assert(MockPausable::is_paused(), 'Should be paused');
```

File: src/openzeppelin/tests/test_pausable.cairo [hunk 8]
```diff
@@ -0,0 +1,46 @@
+use openzeppelin::mocks::mock_pausable::MockPausable;
+
+#[test]
+#[available_gas(2000000)]
+fn test_pause_when_unpaused() {
+    assert(MockPausable::is_paused() == false, 'Should not be paused');
```
Comment (martriay) [inline src/openzeppelin/tests/test_pausable.cairo:6]:
```suggestion
    assert(! MockPausable::is_paused(), 'Should not be paused');
```

File: src/openzeppelin/tests/test_pausable.cairo [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    MockPausable::unpause();
+    MockPausable::assert_unpaused_and_increment();
+    assert(MockPausable::get_count() == 1, 'Should increment');
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('Pausable: not paused', ))]
+fn test_unpause_when_unpaused() {
+    assert(MockPausable::is_paused() == false, 'Should be unpaused');
```
Comment (martriay) [inline src/openzeppelin/tests/test_pausable.cairo:44]:
```suggestion
    assert(! MockPausable::is_paused(), 'Should be unpaused');
```

**Review by martriay — COMMENTED** [review]
looking really good!

File: src/openzeppelin/mocks/mock_pausable.cairo (hunk 2 above)
Comment (andrew-fleming) [reply src/openzeppelin/mocks/mock_pausable.cairo]: Nope. Fixed

Comment (andrew-fleming) [reply src/openzeppelin/mocks/mock_pausable.cairo]: They're not. Fixed

**Review by martriay — APPROVED** [review]
looks good! fix the conflicts and good to merge
